### PR TITLE
[tt-crank] eager SDPA training on the ttml sdpa_fw/sdpa_bw composites

### PR DESCRIPTION
**Pull request by agobeljicTT**
Stacked on #9347. Second of the SDPA training stack (split of #9323): eager path only.

Training through `F.scaled_dot_product_attention` in eager no longer decomposes when the ttml kernels can run it (bf16, 4-D, `Sq == Sk`, `S % 32 == 0`, causal or maskless or one `S x S` bool mask): the choice stub picks the fused op, the forward runs the `sdpa_fw` composite (real logsumexp), the new backward kernel runs `sdpa_bw`. `build_sdpa_fw` / `build_sdpa_bw` own the ttml adaptations (scale folded into Q, lse tile, bool mask rebuild, fully masked rows). DTensor sharding strategy + fake kernel for the backward op. Compiled training keeps MATH until the next PR.

Tests: `test_sdpa.py` eager suites (fused cases incl. GQA / masks / partial grads, decompose fallbacks, dropout, TP/DP multichip). Eager decompose cases are `xfail(strict=False)` for #9344.

### CODE REVIEW

File: tt-crank/python/tt_crank/torch/_sharding.py [hunk 1]
```diff
@@ -79,6 +79,53 @@ def _sdpa_overrideable_sharding(
...
+    max_q,
+    max_k,
+    dropout_p,
+    is_causal,
+    philox_seed,
+    philox_offset,
+    *,
+    scale=None,
+):
+    return torch.empty_like(query), torch.empty_like(key), torch.empty_like(value), None
```
[the remark this answers is not in the document]
Comment (agobeljicTT) [reply tt-crank/python/tt_crank/torch/_sharding.py]: done, the fake returns None for the slots grad_input_mask turns off.

File: tt-crank/tests/python/test_sdpa.py [hunk 2]
```diff
@@ -178,3 +181,206 @@ def record(gm):
...
+    tt_kw = {n: a.to("tt") if isinstance(a, torch.Tensor) else a for n, a in kw.items()}
+    outs: list[torch.Tensor] = []
+
+    def sdpa(a, b, c):
+        outs.append(F.scaled_dot_product_attention(a, b, c, **tt_kw))
+        return outs[-1].sum()
+
+    _run_eager_backward(sdpa, *tts)
+    assert _pcc(outs[-1].detach().cpu(), ref_out) >= pcc, "forward output mismatch"
+    _check_grads(tts, refs, pcc)
```
[the remark this answers is not in the document]
Comment (agobeljicTT) [reply tt-crank/tests/python/test_sdpa.py:294]: added: fused eager cases assert out.grad_fn is ScaledDotProductFusedAttentionOverrideableBackward0, decompose cases assert it is not. the unused constant is gone from this pr, the compile pr uses it.

File: tt-crank/src/engine/ttir_module_builder.cpp [hunk 3]
```diff
@@ -1657,4 +1659,231 @@ mlir::Value build_conv3d(ModuleBuilder &mb, mlir::Value input, mlir::Value weigh
...
+    return 1.0 / std::sqrt(as<double>((d + k_sdpa_tile_width - 1) / k_sdpa_tile_width * k_sdpa_tile_width));
+}
+
+mlir::Value sdpa_fold_scale(ModuleBuilder &mb, mlir::Value query, std::optional<double> scale) {
+    const double alpha = scale.value_or(1.0 / std::sqrt(as<double>(shape_of(query)[3]))) / sdpa_padded_scale(query);
+    return std::fabs(alpha - 1.0) <= 1e-9 ? query : scale_tensor(mb, query, alpha);
+}
+
+// Attention masks, from torch to ttml.
+//
```
Comment (agobeljicTT) [inline tt-crank/src/engine/ttir_module_builder.cpp:1702]: perhaps a highly unnecessary comment, but it does summarize the essence cleanly
@pilkicTT @acolicTT wdy think? perhaps we can remove the whole comment...

File: tt-crank/src/engine/ttir_module_builder.cpp [hunk 4]
```diff
@@ -1868,22 +1885,21 @@ std::tuple<mlir::Value, mlir::Value, mlir::Value> build_sdpa_bw(ModuleBuilder &m
...
-                                             attn_output,
-                                             sdpa_fold_scale(mb, query, scale),
-                                             key,
-                                             value,
-                                             lse};
+    const mlir::Value query_in = sdpa_pad_head_dim(mb, sdpa_fold_scale(mb, query, scale));
+    const mlir::Value key_in = sdpa_pad_head_dim(mb, key);
+    llvm::SmallVector<mlir::Type, 3> result_types{query_in.getType(), key_in.getType(), value.getType()};
+    llvm::SmallVector<mlir::Value, 8> inputs{
+        sdpa_zero_invalid_rows(mb, grad_output, row_valid), attn_output, query_in, key_in, value, lse};
```
Comment (agobeljicTT) [inline tt-crank/src/engine/ttir_module_builder.cpp:1892]: @acolicTT @pilkicTT wdythink do we want to pad this like this or are we comfortable with decompositions?
